File: AoC2023Day05/main.py
def main(input_file: str, part2=False):
    str_maps_to_range_lists = {"seed": [], "soil": [], "fertilizer": [], "water": [],
                               "light": [], "temperature": [], "humidity": []}
    str_maps_traverse_order = ["seed", "soil", "fertilizer", "water", "light", "temperature", "humidity"]
    seeds = []
    location_number = 10000000000000000000000000
    # Read input file and make maps
    with (open(input_file, "r") as file):
        active_map = None
        for line in file.readlines():
            if "seeds:" in line:
                seeds = [int(x) for x in line.strip().split(":")[1].split()]
            else:
                if line == "\n":
                    active_map = None
                    continue
                if active_map is None:
                    for key in str_maps_to_range_lists.keys():
                        if key == line[0:len(key)] and "map" in line:
                            active_map = key
                            break
                else:
                    str_maps_to_range_lists[active_map].append(tuple(map(int, line.strip().split())))

    # print(soil_map)
    for seed in seeds:
        tracked_value = seed  # will ultimately be the location after all maps are traversed
        for key in str_maps_traverse_order:
            for range_tuple in str_maps_to_range_lists[key]:
                if tracked_value in range(range_tuple[1], range_tuple[1] + range_tuple[2]):
                    tracked_value = range_tuple[0] + (tracked_value - range_tuple[1])
                    break
        if tracked_value < location_number:
            location_number = tracked_value
    print(f"Part 1: {location_number}")

    # Part 2
    # seed restructure
    print("Starting part 2")
    part2_seeds = []  # list of tuples (start, length)
    for i in range(0, len(seeds), 2):
        for j in range((seeds[i]), (seeds[i] + seeds[i + 1])):
            part2_seeds.append(j)

    print(f"Part 2 seeds ({len(part2_seeds)}): {part2_seeds}")
    min_loc_value_part_2 = 10000000000000
    for seed in part2_seeds:
        active_seed_traverse_value = seed
        for key in str_maps_traverse_order:
            for range_tuple in str_maps_to_range_lists[key]:
                if active_seed_traverse_value in range(range_tuple[1], range_tuple[1] + range_tuple[2]):
                    active_seed_traverse_value = range_tuple[0] + (active_seed_traverse_value - range_tuple[1])
                    break
        if active_seed_traverse_value < min_loc_value_part_2:
            min_loc_value_part_2 = active_seed_traverse_value

    print(f"Part 2: {min_loc_value_part_2}")


if __name__ == '__main__':
    main('input.txt')
    # main('test.txt')

File: AoC2023Day05/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	inputFile := "input.txt"
	part2 := true
	strMapsToRangeLists := map[string][][3]int64{
		"seed":        {},
		"soil":        {},
		"fertilizer":  {},
		"water":       {},
		"light":       {},
		"temperature": {},
		"humidity":    {},
	}

	strMapsTraverseOrder := []string{"seed", "soil", "fertilizer", "water", "light", "temperature", "humidity"}
	var seeds []int64
	locationNumber := int64(100000000000000)

	file, err := os.Open(inputFile)
	if err != nil {
		fmt.Println("Error opening file:", err)
		return
	}
	defer file.Close()

	activeMap := ""
	scanner := bufio.NewScanner(file)

	for scanner.Scan() {
		line := scanner.Text()
		if strings.Contains(line, "seeds:") {
			seedsStr := strings.Split(line, ":")[1]
			seedTokens := strings.Fields(seedsStr)
			for _, token := range seedTokens {
				seed, _ := strconv.ParseInt(token, 10, 64)
				seeds = append(seeds, seed)
			}
		} else {
			if line == "" {
				activeMap = ""
				continue
			}
			if activeMap == "" {
				for key := range strMapsToRangeLists {
					if strings.HasPrefix(line, key) && strings.Contains(line, "map") {
						activeMap = key
						break
					}
				}
			} else {
				rangeTokens := strings.Fields(line)
				var rangeTuple [3]int64
				for i := 0; i < 3; i++ {
					rangeValue, _ := strconv.ParseInt(rangeTokens[i], 10, 64)
					rangeTuple[i] = rangeValue
				}
				strMapsToRangeLists[activeMap] = append(strMapsToRangeLists[activeMap], rangeTuple)
			}
		}
	}

	for _, seed := range seeds {
		trackedValue := seed
		for _, key := range strMapsTraverseOrder {
			for _, rangeTuple := range strMapsToRangeLists[key] {
				if trackedValue >= rangeTuple[1] && trackedValue < rangeTuple[1]+rangeTuple[2] {
					trackedValue = rangeTuple[0] + (trackedValue - rangeTuple[1])
					break
				}
			}
		}
		if trackedValue < locationNumber {
			locationNumber = trackedValue
		}
	}

	fmt.Printf("Part 1: %d\n", locationNumber)

	if part2 {
		fmt.Println("Starting Part 2")
		traverseListList := make([][][3]int64, len(strMapsTraverseOrder))
		for i, traverseKey := range strMapsTraverseOrder {
			traverseListList[i] = strMapsToRangeLists[traverseKey]
		}
		locationNumber = numbaMethod(seeds, traverseListList)
		fmt.Println("Part 2: ", locationNumber)
	}

}

func numbaMethod(seeds []int64, traverseListList [][][3]int64) int64 {
	minLocValuePart2 := int64(10000000000000)
	for i := 0; i < len(seeds); i += 2 {
		for j := seeds[i]; j < seeds[i]+seeds[i+1]; j++ {
			activeSeedTraverseValue := j

			for _, traverseSet := range traverseListList {
				for _, rangeTuple := range traverseSet {
					if activeSeedTraverseValue >= rangeTuple[1] && activeSeedTraverseValue < rangeTuple[1]+rangeTuple[2] {
						activeSeedTraverseValue = rangeTuple[0] + (activeSeedTraverseValue - rangeTuple[1])
						break
					}
				}
			}
			if activeSeedTraverseValue < minLocValuePart2 {
				minLocValuePart2 = activeSeedTraverseValue
			}
		}
	}
	return minLocValuePart2
}
